handler/src/main/java/io/netty/handler/ssl/SslHandler.java

```diff
@@ -641,7 +646,14 @@ private void wrapNonAppData(ChannelHandlerContext ctx, boolean inUnwrap) throws
                 }
             }
         } catch (SSLException e) {
-            setHandshakeFailure(ctx, e);
+            try {
+                setHandshakeFailure(ctx, e);
+            } finally {
+                // We may have written some parts of data before an exception was thrown so ensure we always flush.
+                // See https://github.com/netty/netty/issues/3900#issuecomment-172481830
+                mayFlush(ctx);
```

Comment:
@Scottmitch I looked again at the code and I think the PR is correct. It will flush() if something was written to the channel  and an exception was thrown (while processing outbound stuff) and on inbound it will flush when channelReadComplete(...) was called and before something was written (which will also handle the case of an exception).